Spokestack/NLUTensorflowSlotParsers.swift

```diff
@@ -0,0 +1,169 @@
+            var slotType = tag
+            if let prefixIndex = tag.range(of: "_")?.upperBound {
+                slotType = String(tag.suffix(from: prefixIndex))
+            }
+            // collect all the tokens, detokenized, with the same slot type into a single array
+            if let tokens = slotTokens[slotType] {
+                if let value = try tokenizer?.detokenize([token]) {
+                    slotTokens[slotType] = tokens + [value]
+                }
+            } else {
+                slotTokens[slotType] = [token]
+            }
+        }
+        // for each tag that isn't unclassified, send the tokens to the slot facet parser
+        for slotName in slotTokens.keys where slotName != "o" {
+            guard let slot = intent.slots.filter({ $0.name == slotName }).first else {
+                throw NLUError.metadata("Could not find a slot called \(slotName) in nlu metadata.")
+            }
+            guard let facetData = slot.facets.data(using: .utf16) else {
+                throw NLUError.metadata("Error when converting \(slotName) facet data.")
+            }
+            let slotValue = try self.slotFacetParser(slot: slot, facetData: facetData, values: slotTokens[slotName])
+
+            slots[slotName] = Slot(type: slot.type, value: slotValue)
+        }
+        
+        return slots
+    }
+    
+    private func slotFacetParser(slot: NLUTensorflowSlot, facetData: Data, values: [String]?) throws -> Any? {
+        guard let text = values else {
+            return nil
+        }
+        switch slot.type {
+        case "selset":
+            let parsedSlot = try decoder.decode(NLUTensorflowSelset.self, from: facetData)
+            // filter the slot selection aliases to see if they match any tokens
+            let contains = parsedSlot.selections.filter { selection in
+                text.contains(where: { name in
+                    selection.aliases.contains(where: { alias in
+                        name == alias
+                    })
+                })
+            }
+            // just pick the first, if any, that matched
+            return contains.first?.name
+        case "integer":
+            let parsedSlot = try decoder.decode(NLUTensorflowInteger.self, from: facetData)
+            guard let lowerBound = parsedSlot.range.first,
+                let upperBound = parsedSlot.range.last,
+                let parsedValue = parseNumber(text)
+                else {
+                return nil
+            }
+            let range = ClosedRange<Int>(uncheckedBounds: (lower: lowerBound, upper: upperBound))
+            return range.contains(parsedValue) ? parsedValue : nil
+        case "digits":
+            let parsedSlot = try decoder.decode(NLUTensorflowDigits.self, from: facetData)
```

Comment:
Not sure what this is trying to point out? The JSON deserialization is done once for each slot.

Reply:
Once for each slot, each time classification runs. What I was saying is that since the metadata is static, it would be best if its deserialization were a one-time cost rather than happening for each utterance that comes in.